README.md
# rk.MultidimensionalScaling

This R package provides a GUI for multidimensional scaling.

![rk.MultidimensionalScaling main tab](./inst/screenshots/main_tab.jpg?raw=true "rk.MultidimensionalScaling main tab")

It is a plugin package for [RKWard](https://rkward.kde.org), a powerful GUI and IDE for [R](https://r-project.org),
and will **only work in conjunction with RKWard**. However, RKWard is free software, please
[check it out](https://rkward.kde.org).

## Installation

### Installation via RKWard

After you installed RKWard, simply install this package like any other R package. The easiest way to get the most
recent stable release is to use RKWard's package management:

- Go to `Settings` -> `Manage R packages and plugins...` -> `Install/Update/Remove R packages`
- Check `Show only packages providing RKWard dialogs` and look at `New Packages`
- Check the desired package(s) and hit `Apply`

After you installed the package, RKWard should immediately have additional menu entries -- there's
usually no need to restart the application.

### Installation via GitHub

To install it directly from GitHub, you can use `install_github()` from the [devtools](https://github.com/hadley/devtools) package:

```
devtools::install_github("rkward-community/rk.MultidimensionalScaling") # stable release
devtools::install_github("rkward-community/rk.MultidimensionalScaling", ref="develop") # development release
```

You can safely ignore warnings about the (unavailable) package "rkward", it was installed together with RKWard.
 
## Contributing

See https://docs.kde.org/trunk5/en/rkward/rkwardplugins/index.html for documentation on writing plugins for RKWard.
Please contact the [RKWard development mailing list](https://mail.kde.org/mailman/listinfo/rkward-devel)
for help, reports and requests.

- Submit pull requests or patches
- Contact us to get write access

This plugin is built using the [rkwarddev](https://files.kde.org/rkward/R/pckg/rkwarddev/index.html) package. This means that its development is mainly focussed on one
script file, located at `rk.MultidimensionalScaling/inst/rkward/rkwarddev_MDS_plugin_script.R`. This script is written in `R` code and *generates*
all of the other plugin files when run. Therefore, in case you'd like to add to this plugin, please add to this script file.

### Branches

Please note that all development happens in the `develop` branch. Pull requests against the `master` branch will be rejected, as it is reserved for the current stable release.

## Licence

Copyright 2012-2022 Meik Michalke <meik.michalke@hhu.de>

rk.MultidimensionalScaling is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

rk.MultidimensionalScaling is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with rk.MultidimensionalScaling.  If not, see <http://www.gnu.org/licenses/>.


ChangeLog
ChangeLog for package rk.MultidimensionalScaling

changes in version 0.2-1 (2022-05-03)
unreleased:
  - this release is under development

changes in version 0.01-12 (2022-05-03)
fixed:
  - screenshot URL and removed flattr-link in README.md
  - replaced deprecated doPrintout() with printout() in JavaScript

changes in version 0.01-11 (2015-12-18)
fixed:
  - corrected variable scope (distance calculation was not correctly enabled/disabled)
  - fixed Changelog version numbering
 added:
  - plugin is now translatable (see inst/rkward/po/)
  - german translation
changed:
  - updated the generator script
  - rebuilt with rkwarddev 0.08-1

changes in version 0.01-10 (2015-06-29)
fixed:
  - fixed issues reading the status of checkboxes in the dialog
  - removed unused "Initial random set of centers" options and made
    the "Maximum number of iterations" options actually work

changes in version 0.01-9 (2015-03-14)
fixed:
  - replaced the outdated links to sourceforge with our new KDE URL
  - missing ChangeLog entry

changes in version 0.01-8 (2015-03-13)
changed:
  - rebuilt using rkwarddev 0.07-1

changes in version 0.01-7 (2013-04-14)
fixed:
  - corrected logic section in main dialog
    (distance calculation was not correctly enabled/disabled)

changes in version 0.01-6 (2013-04-04)
added:
  - debianized the plugin
fixed:
  - removed erroneous pluginmap inclusions from XML files

changes in version 0.01-5 (2013-02-24)
changed:
  - updated rkwarddev generator script (compatibility with 0.06-2) and regenerated plugin code

changes in version 0.01-4 (2012-09-29)
fixed:
  - added generic pre/post plot options

changes in version 0.01-3 (2012-04-07)
changes:
  - added a ChangeLog file
  - rebuild package for R 2.15


inst/rkward/plugins/Multidimensionalscaling.js
// this code was generated using the rkwarddev package.
// perhaps don't make changes here, but in the rkwarddev script instead!

// define variables globally
var frameSelectedVarsChecked;
var selectedVarsShortname;
var frameDataPrepEnabled;

function setGlobalVars(){
  frameSelectedVarsChecked = getValue("frameSelectedVars.checked");
  selectedVarsShortname = getValue("selectedVars.shortname").split("\n").join("\", \"");
  frameDataPrepEnabled = getValue("frameDataPrep.enabled");
}

function preview(){
  preprocess(true);
  calculate(true);
  printout(true);
}

function preprocess(is_preview){
  setGlobalVars();
  // add requirements etc. here
  if(is_preview) {
    echo("if(!base::require(MASS)){stop(" + i18n("Preview not available, because package MASS is not installed or cannot be loaded.") + ")}\n");
  } else {
    echo("require(MASS)\n");
  }
}

function calculate(is_preview){
  // read in variables from dialog
  var data = getString("data");
  var selectedVars = getString("selectedVars");
  var ndim = getString("ndim");
  var distMethod = getString("distMethod");
  var pwrMinkowski = getString("pwrMinkowski");
  var scaleMethod = getString("scaleMethod");
  var maxIter = getString("maxIter");
  var textSize = getString("textSize");
  var textPos = getString("textPos");
  var omitNA = getBoolean("omitNA.state");
  var scale = getBoolean("scale.state");
  var plotResults = getBoolean("plotResults.state");
  var framePlotLabelsChecked = getBoolean("framePlotLabels.checked");

  // the R code to be evaluated
  var frameSelectedVarsChecked = getValue("frameSelectedVars.checked");
  var selectedVarsShortname = getValue("selectedVars.shortname").split("\n").join("\", \"");
  var frameDataPrepEnabled = getValue("frameDataPrep.enabled");
  if(frameSelectedVarsChecked && selectedVarsShortname != "") {
    comment("Use subset of variables", "  ");  
    echo("\t" + data + " <- subset(" + data + ", select=c(\"" + selectedVarsShortname + "\"))\n");  
  } else {}
  if(frameDataPrepEnabled && omitNA) {
    comment("Listwise removal of missings", "  ");  
    echo("\t" + data + " <- na.omit(" + data + ")\n");  
  } else {}
  if(frameDataPrepEnabled && scale) {
    comment("Standardizing values", "  ");  
    echo("\t" + data + " <- scale(" + data + ")\n");  
  } else {}
  if(frameDataPrepEnabled) {
    comment("Compute distance matrix", "  ");  
    echo("\tmds.distances <- dist(");  
    if(data) {
      echo("\n\t\tx=" + data);  
    } else {}  
    echo(",\n\t\tmethod=\"" + distMethod + "\"");  
    if(distMethod == "minkowski") {
      echo(",\n\t\tp=" + pwrMinkowski);  
    } else {}  
    echo("\n\t)\n");  
    comment("The actual multidimensional scaling", "  ");  
    echo("\tmds.result <- " + scaleMethod + "(");  
    if(data) {
      echo("\n\t\td=mds.distances");  
    } else {}  
    echo(",\n\t\tk=" + ndim);  
    if(scaleMethod == "isoMDS") {
      echo(",\n\t\tmaxit=" + maxIter);  
    } else if(scaleMethod == "sammon") {
      echo(",\n\t\tniter=" + maxIter);  
    } else {}  
    echo("\n\t)\n\n");  
  } else {
    comment("The actual multidimensional scaling", "  ");  
    echo("\tmds.result <- " + scaleMethod + "(");  
    if(data) {
      echo("\n\t\td=" + data);  
    } else {}  
    echo(",\n\t\tk=" + ndim);  
    if(scaleMethod == "isoMDS") {
      echo(",\n\t\tmaxit=" + maxIter);  
    } else if(scaleMethod == "sammon") {
      echo(",\n\t\tniter=" + maxIter);  
    } else {}  
    echo("\n\t)\n\n");  
  }
}

function printout(is_preview){
  // read in variables from dialog
  var data = getString("data");
  var selectedVars = getString("selectedVars");
  var ndim = getString("ndim");
  var distMethod = getString("distMethod");
  var pwrMinkowski = getString("pwrMinkowski");
  var scaleMethod = getString("scaleMethod");
  var maxIter = getString("maxIter");
  var textSize = getString("textSize");
  var textPos = getString("textPos");
  var omitNA = getBoolean("omitNA.state");
  var scale = getBoolean("scale.state");
  var plotResults = getBoolean("plotResults.state");
  var framePlotLabelsChecked = getBoolean("framePlotLabels.checked");

  // printout the results
  if(!is_preview) {
    new Header(i18n("Multidimensional scaling")).print();  
  } else {}  var framePlotLabelsChecked = getValue("framePlotLabels.checked");
  if(plotResults) {
    echo("\n");  
    var textColCodePrintout = getValue("textCol.code.printout");  
        // in case there are generic plot options defined:
    var genPlotOptionsCodePreprocess = getValue("genPlotOptions.code.preprocess");
    var genPlotOptionsCodePrintout = getValue("genPlotOptions.code.printout");
    var genPlotOptionsCodeCalculate = getValue("genPlotOptions.code.calculate");

    if(!is_preview) {
    echo("rk.graph.on()\n");  
  } else {}
    echo("    try({\n");

    // insert any option-setting code that should be run before the actual plotting commands:
    printIndentedUnlessEmpty("      ", genPlotOptionsCodePreprocess, "\n", "");

    // the actual plot:
    // label text color:
    echo("\t\tplot(mds.result");
    if(scaleMethod == "isoMDS" || scaleMethod == "sammon") {
      echo("[[\"points\"]]");  
    } else {}
    if(!genPlotOptionsCodePrintout.match(/main\s*=/)) {
      echo(",\n\t\t\tmain=\"Multidimensional scaling\"");  
    } else {}
    if(!genPlotOptionsCodePrintout.match(/sub\s*=/)) {
      echo(",\n\t\t\tsub=\"Solution with " + ndim + " dimensions (" + scaleMethod + ")\"");  
    } else {}
    if(framePlotLabelsChecked && textPos == 0) {
      echo(",\n\t\t\ttype=\"n\"");  
    } else {}
    echo(genPlotOptionsCodePrintout.replace(/, /g, ",\n\t\t\t"));
    echo(")");
    if(framePlotLabelsChecked) {
      echo("\n\t\ttext(mds.result");  
      if(scaleMethod == "isoMDS" || scaleMethod == "sammon") {
        echo("[[\"points\"]],\n\t\t\trownames(mds.result[[\"points\"]])");  
      } else {
        echo(",\n\t\t\trownames(mds.result)");  
      }  
      if(textSize != 1) {
        echo(",\n\t\t\tcex=" + textSize);  
      } else {}  
      if(textPos != 0) {
        echo(",\n\t\t\tpos=" + textPos);  
      } else {}  
      echo(textColCodePrintout + ")");  
    } else {}

    // insert any option-setting code that should be run after the actual plot:
    printIndentedUnlessEmpty("      ", genPlotOptionsCodeCalculate, "\n", "");

    echo("\n    })\n");
    if(!is_preview) {
    echo("rk.graph.off()\n");  
  } else {}  
  } else {}
  if(!is_preview) {
    echo("\nrk.print(mds.result)\n");  
    if(frameSelectedVarsChecked && selectedVarsShortname != "") {
      new Header(i18n("Subset of variables included the analysis"), 3).print();  
      echo("rk.print(list(\"" + selectedVarsShortname + "\"))\n\n");  
    } else {}  
  } else {}
  if(!is_preview) {
    //// save result object
    // read in saveobject variables
    var saveResults = getValue("saveResults");
    var saveResultsActive = getValue("saveResults.active");
    var saveResultsParent = getValue("saveResults.parent");
    // assign object to chosen environment
    if(saveResultsActive) {
      echo(".GlobalEnv$" + saveResults + " <- mds.result\n");
    } else {}  
  } else {}

}



inst/rkward/rkwarddev_MDS_plugin_script.R
# the plugin code was generated by this script
# you should not change the plugin code directly, but this script
# note: this script only creates objects in your workspace,
# *EXCEPT* for the last call, see below.

require(rkwarddev)
rkwarddev.required("0.08-1")

local({
# set the output directory to overwrite the actual plugin
output.dir <- tempdir()
overwrite <- TRUE
# if you set guess.getters to TRUE, the resulting code will need RKWard >= 0.6.0
guess.getter <- TRUE
rk.set.indent(by="  ")
rk.set.empty.e(TRUE)
update.translations <- TRUE

about.info <- rk.XML.about(
  name="rk.MultidimensionalScaling",
  author=c(
    person(given="Meik", family="Michalke",
      email="meik.michalke@hhu.de", role=c("aut","cre"))),
  about=list(desc="RKWard GUI for multidimensional scaling",
    version="0.2-1", url="https://rkward.kde.org")
  )
dependencies.info <- rk.XML.dependencies(
  dependencies=list(rkward.min=ifelse(isTRUE(guess.getter), "0.6.0", "0.5.6")),
  package=list(c(name="MASS"))
)

############
## prepare data
############
omitNA <- rk.XML.cbox("Remove missing values", value="true", chk=TRUE, id.name="omitNA")
scale <- rk.XML.cbox("Stadardize values", value="true", id.name="scale")
frameDataPrep <- rk.XML.frame(
  rk.XML.row(
    rk.XML.col(omitNA),
    rk.XML.col(scale)
  ),
  label="Data preparation",
  id.name="frameDataPrep")

genPlotOptions <- rk.plotOptions(id.name="genPlotOptions")
textCol <- rk.plotOptions(embed="rkward::color_chooser", button=FALSE, id.name="textCol")

############
## classical multidimensional scaling
############
varSelect <- rk.XML.varselector(label="Select data", id.name="varSelect")
data <- rk.XML.varslot(label="Data (data.frame, matrix or dist)", source=varSelect, classes=c("data.frame", "matrix", "dist"), required=TRUE, id.name="data")
selectedVars <- rk.XML.varslot(label="Selected variables", source=varSelect, multi=TRUE, id.name="selectedVars")
frameSelectedVars <- rk.XML.frame(selectedVars, label="Use only a subset of variables", checkable=TRUE, chk=FALSE, id.name="frameSelectedVars")

ndim <- rk.XML.spinbox(label="Maximum dimensions", min=2, real=FALSE, id.name="ndim")

scaleMethod <- rk.XML.dropdown(label="Scaling method", options=list(
    "Classical (metric)"=c(val="cmdscale", chk=TRUE),
    "Kruskal (non-metric)"=c(val="isoMDS"),
    "Sammon (non-linear)"=c(val="sammon")
  ),
  id.name="scaleMethod"
)

distMethod <- rk.XML.dropdown(label="Computation method", options=list(
    "Euclidean"=c(val="euclidean", chk=TRUE),
    "Maximum"=c(val="maximum"),
    "Manhattan (city block)"=c(val="manhattan"),
    "Canberra"=c(val="canberra"),
    "Binary"=c(val="binary"),
    "Minkowski"=c(val="minkowski")
  ),
  id.name="distMethod"
)
pwrMinkowski <- rk.XML.spinbox(label="Power of Minkowski distance", min=1, initial=2, real=FALSE, id.name="pwrMinkowski")

maxIter <- rk.XML.spinbox(label="Maximum number of iterations", min=1, initial=50, real=FALSE, id.name="maxIter")
#mds.spin.nstart <- rk.XML.spinbox(label="Initial random set of centers", min=1, initial=1, real=FALSE)

saveResults <- rk.XML.saveobj("Save results to workspace", initial="mds.result", id.name="saveResults")

# plot results
plotResults <- rk.XML.cbox("Plot results", val="true", chk=TRUE, id.name="plotResults")
textSize <- rk.XML.spinbox("Text size", initial=0.8, id.name="textSize")
textPos <- rk.XML.dropdown(label="Text position", options=list(
    "Instead of point"=c(val=0),
    "Below point"=c(val=1, chk=TRUE),
    "Left to point"=c(val=2),
    "Above point"=c(val=3),
    "Right to point"=c(val=4)
  ),
  id.name="textPos"
)
framePlotLabels <- rk.XML.frame(
  textCol,
  textSize,
  textPos,
  label="Plot labels (from row names of data)",
  checkable=TRUE,
  chk=TRUE,
  id.name="framePlotLabels"
)

mds.plot.preview <- rk.XML.preview()

tab.mds.data <- rk.XML.row(
    varSelect,
    rk.XML.col(
      data,
      frameSelectedVars,
      frameDataPrep,
      rk.XML.stretch(),
      saveResults
    ),
    rk.XML.col(
      rk.XML.frame(ndim),
      rk.XML.stretch(),
      mds.frame.dist <- rk.XML.frame(distMethod, pwrMinkowski, label="Distance matrix"),
      rk.XML.frame(
        scaleMethod,
        rk.XML.row(
          maxIter
#           rk.XML.col(maxIter),
#           rk.XML.col(mds.spin.nstart)
        ),
        plotResults,
        label="Advanced options")
    )
  )

tab.mds.plot <- rk.XML.row(
    rk.XML.col(
      framePlotLabels,
      rk.XML.stretch(),
      genPlotOptions,
      mds.plot.preview)
  )

mds.full.dialog <- rk.XML.dialog(
  rk.XML.tabbook(tabs=list(
    "Data"=tab.mds.data,
    "Plot"=tab.mds.plot
  )),
  label="Multidimensional scaling")

lgc.sect.mds <- rk.XML.logic(
    lgc.current.object <- rk.XML.connect(governor="current_object", client=data, set="available"),
    lgc.data.from.selection <- rk.XML.connect(governor=data, client=varSelect, get="available", set="root"),
    gov.data <- rk.XML.convert(sources=list(available=data), mode=c(notequals="")),
    lgc.enable.selected <- rk.XML.connect(governor=gov.data, client=frameSelectedVars, set="enabled"),
  lgc.df.script <- rk.comment(id("
    gui.addChangeCommand(", idq(data, modifiers="available", js=FALSE), ", \"dataChanged()\");
    // this function is called whenever the data was changed
    dataChanged = function(){
        var prepareFrame = \"true\";
        var selectFrame = \"true\";
        var thisObject = makeRObject(gui.getValue(", idq(data, modifiers="available", js=FALSE), "));
        if(thisObject.classes()){
          if(!thisObject.isDataFrame()){
            selectFrame = \"false\";
            if(thisObject.classes().indexOf(\"dist\") != -1){
              prepareFrame = \"false\";
            } else {}
          } else {}
        } else {}
        gui.setValue(", idq(frameSelectedVars, modifiers="enabled", js=FALSE), ", selectFrame);
        gui.setValue(", idq(frameDataPrep, modifiers="enabled", js=FALSE), ", prepareFrame);
      }", js=FALSE)),
  MDS.gov.dist <- rk.XML.convert(sources=list(string=distMethod), mode=c(equals="minkowski")),
  rk.XML.connect(governor=MDS.gov.dist, client=pwrMinkowski, set="enabled"),
  rk.XML.connect(governor=plotResults, client=tab.mds.plot, set="enabled"),
  MDS.gov.meth <- rk.XML.convert(sources=list(string=scaleMethod), mode=c(notequals="cmdscale")),
  rk.XML.connect(governor=MDS.gov.meth, client=maxIter, set="enabled"),
  # disable distance computation, if dist object given
  lgc.isntDistData <- rk.XML.connect(governor=frameDataPrep, get="enabled", client=mds.frame.dist, set="enabled"),
  # set label text color to red
  rk.XML.set(textCol, set="color.string", to="red")
)

## JavaScript
js.frm.subset <- rk.JS.vars(frameSelectedVars, modifiers="checked") # see if the frame is checked
js.selectedVars <- rk.JS.vars(selectedVars, modifiers="shortname", join="\\\", \\\"") # get selected vars
js.prepare <- rk.JS.vars(frameDataPrep, modifiers="enabled") # see if data preparation is off

js.global.vars <- list(js.frm.subset, js.selectedVars, js.prepare)

mds.js.calc <- rk.paste.JS(
  js.frm.subset, # see if the frame is checked
  js.selectedVars, # get selected vars
  js.prepare, # see if data preparation is off
  js(
    if(js.frm.subset && js.selectedVars != ""){
      R.comment("Use subset of variables")
      echo("\t", data, " <- subset(",data,", select=c(\"", js.selectedVars, "\"))\n")
    },
    if(js.prepare && omitNA){
      R.comment("Listwise removal of missings")
      echo("\t", data, " <- na.omit(", data, ")\n")
    },
    if(js.prepare && scale){
      R.comment("Standardizing values")
      echo("\t", data, " <- scale(", data, ")\n")
    }
  ),
  js(
    if(js.prepare){
      R.comment("Compute distance matrix")
      echo("\tmds.distances <- dist(")
      if(data){
        echo("\n\t\tx=", data)
      } else {}
      echo(",\n\t\tmethod=\"", distMethod, "\"")
      if(distMethod == "minkowski"){
        echo(",\n\t\tp=", pwrMinkowski)
      } else {}
      echo("\n\t)\n")
      R.comment("The actual multidimensional scaling")
      echo("\tmds.result <- ", scaleMethod,"(")
      if(data){
        echo("\n\t\td=mds.distances")
      } else {}
      echo(",\n\t\tk=", ndim)
      if(scaleMethod == "isoMDS"){
        echo(",\n\t\tmaxit=", maxIter)
      } else if(scaleMethod == "sammon"){
          echo(",\n\t\tniter=", maxIter)
      } else {}
      echo("\n\t)\n\n")
    } else {
      R.comment("The actual multidimensional scaling")
      echo("\tmds.result <- ", scaleMethod,"(")
      if(data){
        echo("\n\t\td=", data)
      }
      echo(",\n\t\tk=", ndim)
      if(scaleMethod == "isoMDS"){
        echo(",\n\t\tmaxit=", maxIter)
      } else if(scaleMethod == "sammon"){
        echo(",\n\t\tniter=", maxIter)
      } else {}
      echo("\n\t)\n\n")
    }
  )
)

mds.js.plot <- rk.paste.JS(
  js.frm.labels <- rk.JS.vars(framePlotLabels, modifiers="checked"),
  js(
    if(plotResults){
      echo("\n")
      rk.paste.JS(textCol, level=1)
      rk.paste.JS.graph(
        rk.comment("label text color:"),
        echo("\t\tplot(mds.result"),
        js(
          if(scaleMethod == "isoMDS" || scaleMethod == "sammon"){
            echo("[[\"points\"]]")
          } else {},
          if(id("!", genPlotOptions, ".match(/main\\s*=/)")){
            echo(",\n\t\t\tmain=\"Multidimensional scaling\"")
          } else {},
          if(id("!", genPlotOptions, ".match(/sub\\s*=/)")){
            echo(",\n\t\t\tsub=\"Solution with ", ndim, " dimensions (", scaleMethod, ")\"")
          } else {},
          # turn off points if labels should replace them
          if(js.frm.labels && textPos == 0){
            echo(",\n\t\t\ttype=\"n\"")
          } else {},
          # generic plot options go here
          id("echo(", genPlotOptions, ".replace(/, /g, \",\\n\\t\\t\\t\"));"),
          echo(")"),
          if(js.frm.labels){
            echo("\n\t\ttext(mds.result")
            if(scaleMethod == "isoMDS" || scaleMethod == "sammon"){
              echo("[[\"points\"]],\n\t\t\trownames(mds.result[[\"points\"]])")
            } else {
              echo(",\n\t\t\trownames(mds.result)")
            }
            if(textSize != 1){
              echo(",\n\t\t\tcex=", textSize)
            } else {}
            if(textPos != 0){
              echo(",\n\t\t\tpos=", textPos)
            } else {}
            echo(textCol, ")")
          } else {},
          level=3
        ),
        plotOpts=genPlotOptions,
        level=3
      )
    } else {},
    if("!is_preview"){
      echo("\nrk.print(mds.result)\n")
      # print selected subsets, if needed
        if(js.frm.subset && js.selectedVars != ""){
          rk.JS.header("Subset of variables included the analysis", level=3)
          echo("rk.print(list(\"", js.selectedVars, "\"))\n\n")
        } else {}
    } else {}
  )
)



#############
## if you run the following function call, files will be written to tempdir!
#############
# this is where it get's serious, that is, here all of the above is put together into one plugin

mds.plugin.dir <<- rk.plugin.skeleton(
  about.info,
  path=output.dir,
  guess.getter=guess.getter,
  xml=list(
    dialog=mds.full.dialog,
    logic=lgc.sect.mds),
  js=list(results.header="Multidimensional scaling",
    globals=js.global.vars,
    require="MASS",
    calculate=mds.js.calc,
    printout=mds.js.plot),
  pluginmap=list(name="Multidimensional scaling", hierarchy=list("analysis")),
  dependencies=dependencies.info,
  create=c("pmap", "xml", "js", "desc"),
  overwrite=overwrite,
  tests=FALSE,
#  edit=TRUE,
  load=TRUE,
#  show=TRUE,
  hints=FALSE)

  if(isTRUE(update.translations)){
    rk.updatePluginMessages(file.path(output.dir,"rk.MultidimensionalScaling","inst","rkward","rk.MultidimensionalScaling.pluginmap"))
  } else {}
})
